File: shared/src/commonMain/kotlin/org/nihongo/mochi/ui/components/RecapComponents.kt
package org.nihongo.mochi.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.luminance
import org.jetbrains.compose.resources.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.nihongo.mochi.domain.kanji.KanjiEntry
import org.nihongo.mochi.shared.generated.resources.Res
import org.nihongo.mochi.shared.generated.resources.game_recap_play
import org.nihongo.mochi.shared.generated.resources.mode_revise
import org.nihongo.mochi.shared.generated.resources.previous_page
import org.nihongo.mochi.shared.generated.resources.next_page
import org.nihongo.mochi.shared.generated.resources.order_by
import org.nihongo.mochi.shared.generated.resources.size

@Composable
fun RecapKanjiGrid(
    kanjiList: List<Pair<KanjiEntry, Color>>,
    onKanjiClick: (KanjiEntry) -> Unit
) {
    LazyVerticalGrid(
        columns = GridCells.Fixed(8),
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(4.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp)
    ) {
        items(kanjiList) { (kanjiEntry, color) ->
            RecapKanjiGridItem(
                kanji = kanjiEntry.character,
                color = color,
                onClick = { onKanjiClick(kanjiEntry) }
            )
        }
    }
}

@Composable
fun RecapKanjiGridItem(
    kanji: String,
    color: Color,
    onClick: () -> Unit
) {
    // Determine a readable text color based on background luminance
    val textColor = if (color.luminance() > 0.5f) Color.Black else Color.White

    Box(
        modifier = Modifier
            .clip(RoundedCornerShape(4.dp))
            .background(color)
            .aspectRatio(1f) // Ensure square cells
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = kanji,
            fontSize = 24.sp,
            color = textColor,
            textAlign = TextAlign.Center
        )
    }
}

@Composable
fun PaginationControls(
    currentPage: Int,
    totalPages: Int,
    onPrevClick: () -> Unit,
    onNextClick: () -> Unit
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center
    ) {
        IconButton(onClick = onPrevClick, enabled = currentPage > 0) {
            Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = stringResource(Res.string.previous_page), tint = MaterialTheme.colorScheme.onBackground)
        }
        Text(
            text = "Page ${currentPage + 1} of $totalPages",
            modifier = Modifier.padding(horizontal = 16.dp),
            textAlign = TextAlign.Center,
            color = MaterialTheme.colorScheme.onBackground
        )
        IconButton(onClick = onNextClick, enabled = currentPage < totalPages - 1) {
            Icon(Icons.AutoMirrored.Filled.ArrowForward, contentDescription = stringResource(Res.string.next_page), tint = MaterialTheme.colorScheme.onBackground)
        }
    }
}

@Composable
fun <T> ModeSelector(
    title: String? = null,
    options: List<Pair<String, T>>,
    selectedOption: T,
    onOptionSelected: (T) -> Unit,
    enabled: Boolean = true
) {
    Column(modifier = Modifier.fillMaxWidth(), horizontalAlignment = Alignment.CenterHorizontally) {
        title?.let {
            Text(text = it, style = MaterialTheme.typography.titleSmall, modifier = Modifier.padding(bottom = 4.dp), color = MaterialTheme.colorScheme.onBackground)
        }
        
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 8.dp)
                .clip(RoundedCornerShape(12.dp))
                .background(MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.8f))
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth()
            ) {
                options.forEach { (text, value) ->
                    Row(
                        modifier = Modifier
                            .weight(1f) // Give equal space to each option
                            .clip(RoundedCornerShape(8.dp))
                            .clickable(enabled = enabled) { onOptionSelected(value) }
                            .padding(vertical = 4.dp, horizontal = 8.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.Center // Center content inside the weighted space
                    ) {
                        RadioButton(
                            selected = (value == selectedOption),
                            onClick = { onOptionSelected(value) },
                            enabled = enabled,
                            colors = RadioButtonDefaults.colors(selectedColor = MaterialTheme.colorScheme.primary)
                        )
                        Text(
                            text = text, 
                            color = if (value == selectedOption) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant, 
                            fontWeight = FontWeight.Bold
                        )
                    }
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun <T> DropdownSelector(
    title: String? = null,
    options: List<Pair<String, T>>,
    selectedOption: T,
    onOptionSelected: (T) -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true
) {
    var expanded by remember { mutableStateOf(false) }
    val selectedText = options.find { it.second == selectedOption }?.first ?: ""

    Column(
        modifier = modifier.padding(vertical = 4.dp),
        horizontalAlignment = Alignment.Start
    ) {
        title?.let {
            Text(
                text = it,
                style = MaterialTheme.typography.titleSmall,
                modifier = Modifier.padding(bottom = 4.dp),
                color = MaterialTheme.colorScheme.onBackground
            )
        }

        Text(
            text = stringResource(Res.string.order_by),
            style = MaterialTheme.typography.titleSmall,
            modifier = Modifier.padding(bottom = 4.dp),
            color = MaterialTheme.colorScheme.onBackground
        )

        ExposedDropdownMenuBox(
            expanded = expanded,
            onExpandedChange = { if (enabled) expanded = !expanded },
            modifier = Modifier.fillMaxWidth()
        ) {
            OutlinedTextField(
                value = selectedText,
                onValueChange = {},
                readOnly = true,
                enabled = enabled,
                trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
                colors = OutlinedTextFieldDefaults.colors(
                    focusedBorderColor = MaterialTheme.colorScheme.primary,
                    unfocusedBorderColor = MaterialTheme.colorScheme.primary,
                    focusedContainerColor = MaterialTheme.colorScheme.primary.copy(alpha = 0.7f),
                    unfocusedContainerColor = MaterialTheme.colorScheme.primary.copy(alpha = 0.7f),
                ),
                modifier = Modifier
                    .menuAnchor(MenuAnchorType.PrimaryNotEditable, enabled)
                    .fillMaxWidth(),
                shape = RoundedCornerShape(12.dp)
            )

            ExposedDropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false }
            ) {
                options.forEach { (text, value) ->
                    DropdownMenuItem(
                        text = { Text(text = text) },
                        onClick = {
                            onOptionSelected(value)
                            expanded = false
                        }
                    )
                }
            }
        }
    }
}

@Composable
fun QuizSizeInput(
    size: String,
    onSizeChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true
) {
    Column(
        modifier = modifier.padding(vertical = 4.dp),
        horizontalAlignment = Alignment.Start
    ) {
        Text(
            text = stringResource(Res.string.size),
            style = MaterialTheme.typography.titleSmall,
            modifier = Modifier.padding(bottom = 4.dp),
            color = MaterialTheme.colorScheme.onBackground
        )
        
        OutlinedTextField(
            value = size,
            onValueChange = { if (it.length <= 3 && it.all { char -> char.isDigit() }) onSizeChange(it) },
            enabled = enabled,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            colors = OutlinedTextFieldDefaults.colors(
                focusedBorderColor = MaterialTheme.colorScheme.primary,
                unfocusedBorderColor = MaterialTheme.colorScheme.primary,
                focusedContainerColor = MaterialTheme.colorScheme.primary.copy(alpha = 0.7f),
                unfocusedContainerColor = MaterialTheme.colorScheme.primary.copy(alpha = 0.7f),
                ),
            modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(12.dp),
            singleLine = true
        )
    }
}

@Composable
fun PlayAndReviewButtons(
    onPlayClick: () -> Unit,
    onReviewClick: () -> Unit,
    isReviewEnabled: Boolean,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(120.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Button(
            onClick = onPlayClick,
            modifier = Modifier
                .weight(1f)
                .fillMaxHeight(),
            elevation = ButtonDefaults.buttonElevation(defaultElevation = 16.dp),
            colors = ButtonDefaults.buttonColors(
                containerColor = MaterialTheme.colorScheme.primary,
                contentColor = MaterialTheme.colorScheme.onPrimary
            ),
            shape = RoundedCornerShape(16.dp)
        ) {
            Text(
                text = stringResource(Res.string.game_recap_play),
                fontSize = 32.sp,
                fontWeight = FontWeight.Bold
            )
        }
        
        Button(
            onClick = onReviewClick,
            enabled = isReviewEnabled,
            modifier = Modifier
                .weight(1f)
                .fillMaxHeight(),
            elevation = ButtonDefaults.buttonElevation(defaultElevation = 16.dp),
            colors = ButtonDefaults.buttonColors(
                containerColor = MaterialTheme.colorScheme.primary,
                contentColor = MaterialTheme.colorScheme.onPrimary,
                disabledContainerColor = MaterialTheme.colorScheme.secondary.copy(alpha = 0.3f),
                disabledContentColor = MaterialTheme.colorScheme.onSecondary.copy(alpha = 0.3f)
            ),
            shape = RoundedCornerShape(16.dp)
        ) {
            Text(
                text = stringResource(Res.string.mode_revise),
                fontSize = 32.sp,
                fontWeight = FontWeight.Bold
            )
        }
    }
}

@Composable
fun PlayButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Button(
        onClick = onClick,
        modifier = modifier
            .fillMaxWidth()
            .height(120.dp),
        elevation = ButtonDefaults.buttonElevation(defaultElevation = 16.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = MaterialTheme.colorScheme.primary,
            contentColor = MaterialTheme.colorScheme.onPrimary
        )
    ) {
        Text(
            text = stringResource(Res.string.game_recap_play),
            fontSize = 48.sp,
            fontWeight = FontWeight.Bold
        )
    }
}

File: flutter_app/lib/widgets/recap_components.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../providers/settings_provider.dart';

class RecapGridItem extends StatelessWidget {
  final String character;
  final Color color;
  final VoidCallback onClick;

  const RecapGridItem({
    super.key,
    required this.character,
    required this.color,
    required this.onClick,
  });

  @override
  Widget build(BuildContext context) {
    final textColor = color.computeLuminance() > 0.5 ? Colors.black : Colors.white;

    return InkWell(
      onTap: onClick,
      child: Container(
        decoration: BoxDecoration(
          color: color,
          borderRadius: BorderRadius.circular(4),
        ),
        alignment: Alignment.center,
        child: LayoutBuilder(
          builder: (context, constraints) {
            final fontSize = (constraints.maxWidth * 0.45).clamp(18.0, 36.0);
            return Text(
              character,
              style: TextStyle(
                fontSize: fontSize,
                color: textColor,
                fontWeight: FontWeight.bold,
              ),
            );
          },
        ),
      ),
    );
  }
}

class PaginationControls extends StatelessWidget {
  final int currentPage;
  final int totalPages;
  final VoidCallback onPrevClick;
  final VoidCallback onNextClick;

  const PaginationControls({
    super.key,
    required this.currentPage,
    required this.totalPages,
    required this.onPrevClick,
    required this.onNextClick,
  });

  @override
  Widget build(BuildContext context) {
    final onBackground = Theme.of(context).colorScheme.onSurface;

    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        IconButton(
          onPressed: currentPage > 0 ? onPrevClick : null,
          icon: Icon(Icons.arrow_back, color: onBackground),
        ),
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16),
          child: Text(
            "Page ${currentPage + 1} of $totalPages",
            style: TextStyle(fontWeight: FontWeight.bold, color: onBackground),
          ),
        ),
        IconButton(
          onPressed: currentPage < totalPages - 1 ? onNextClick : null,
          icon: Icon(Icons.arrow_forward, color: onBackground),
        ),
      ],
    );
  }
}

class ModeSelector<T> extends StatelessWidget {
  final String? title;
  final List<MapEntry<String, T>> options;
  final T selectedOption;
  final ValueChanged<T> onOptionSelected;
  final bool enabled;

  const ModeSelector({
    super.key,
    this.title,
    required this.options,
    required this.selectedOption,
    required this.onOptionSelected,
    this.enabled = true,
  });

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return Column(
      children: [
        if (title != null)
          Padding(
            padding: const EdgeInsets.only(bottom: 4.0),
            child: Text(
              title!,
              style: TextStyle(fontSize: 12, fontWeight: FontWeight.bold, color: theme.colorScheme.onSurface.withValues(alpha: 0.7)),
            ),
          ),
        Container(
          margin: const EdgeInsets.symmetric(horizontal: 8),
          decoration: BoxDecoration(
            color: theme.colorScheme.surfaceContainerHighest.withValues(alpha: 0.8),
            borderRadius: BorderRadius.circular(12),
          ),
          child: Row(
            children: options.map((entry) {
              final isSelected = entry.value == selectedOption;
              return Expanded(
                child: InkWell(
                  onTap: enabled ? () => onOptionSelected(entry.value) : null,
                  borderRadius: BorderRadius.circular(12),
                  child: Container(
                    padding: const EdgeInsets.symmetric(vertical: 8),
                    decoration: isSelected
                        ? BoxDecoration(
                            color: theme.colorScheme.primary.withValues(alpha: 0.2),
                            borderRadius: BorderRadius.circular(12),
                            border: Border.all(color: theme.colorScheme.primary, width: 2),
                          )
                        : null,
                    child: Text(
                      entry.key,
                      textAlign: TextAlign.center,
                      style: TextStyle(
                        fontWeight: FontWeight.bold,
                        color: isSelected ? theme.colorScheme.primary : theme.colorScheme.onSurfaceVariant,
                      ),
                    ),
                  ),
                ),
              );
            }).toList(),
          ),
        ),
      ],
    );
  }
}

class PlayAndReviewButtons extends StatelessWidget {
  final VoidCallback onPlayClick;
  final VoidCallback onReviewClick;
  final bool isReviewEnabled;

  const PlayAndReviewButtons({
    super.key,
    required this.onPlayClick,
    required this.onReviewClick,
    required this.isReviewEnabled,
  });

  @override
  Widget build(BuildContext context) {
    final settings = context.read<SettingsProvider>();
    final theme = Theme.of(context);

    return Row(
      children: [
        Expanded(
          child: _BigButton(
            label: settings.getString("game_recap_play").toUpperCase(),
            onPressed: onPlayClick,
            backgroundColor: theme.colorScheme.primary,
            foregroundColor: theme.colorScheme.onPrimary,
          ),
        ),
        const SizedBox(width: 12),
        Expanded(
          child: _BigButton(
            label: settings.getString("mode_revise").toUpperCase(),
            onPressed: isReviewEnabled ? onReviewClick : null,
            backgroundColor: theme.colorScheme.primary,
            foregroundColor: theme.colorScheme.onPrimary,
            disabledBackgroundColor: theme.colorScheme.secondary.withValues(alpha: 0.3),
            disabledForegroundColor: theme.colorScheme.onSecondary.withValues(alpha: 0.3),
          ),
        ),
      ],
    );
  }
}

class PlayButton extends StatelessWidget {
  final VoidCallback onClick;

  const PlayButton({super.key, required this.onClick});

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final settings = context.read<SettingsProvider>();

    return ElevatedButton(
      onPressed: onClick,
      style: ElevatedButton.styleFrom(
        backgroundColor: theme.colorScheme.primary,
        foregroundColor: theme.colorScheme.onPrimary,
        minimumSize: const Size.fromHeight(120),
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
        elevation: 16,
      ),
      child: Text(
        settings.getString("game_recap_play").toUpperCase(),
        style: const TextStyle(fontSize: 48, fontWeight: FontWeight.bold),
      ),
    );
  }
}

class _BigButton extends StatelessWidget {
  final String label;
  final VoidCallback? onPressed;
  final Color backgroundColor;
  final Color foregroundColor;
  final Color? disabledBackgroundColor;
  final Color? disabledForegroundColor;

  const _BigButton({
    required this.label,
    required this.onPressed,
    required this.backgroundColor,
    required this.foregroundColor,
    this.disabledBackgroundColor,
    this.disabledForegroundColor,
  });

  @override
  Widget build(BuildContext context) {
    return ElevatedButton(
      onPressed: onPressed,
      style: ElevatedButton.styleFrom(
        backgroundColor: backgroundColor,
        foregroundColor: foregroundColor,
        disabledBackgroundColor: disabledBackgroundColor ?? backgroundColor.withValues(alpha: 0.3),
        disabledForegroundColor: disabledForegroundColor ?? foregroundColor.withValues(alpha: 0.3),
        minimumSize: const Size.fromHeight(120),
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
        elevation: 16,
      ),
      child: Text(
        label,
        style: const TextStyle(fontSize: 32, fontWeight: FontWeight.bold),
      ),
    );
  }
}

